Cycling Analysis/pace_curve.py

```diff
@@ -0,0 +1,148 @@
+            return f"{seconds / 3600:.1f} hr"
+
+
+    def create_pace_curve(self, activity_type, date, num_days):
+        """
+        Creates a pace curve based on the specified activity type, date, and number of days.
+
+        Args:
+            activity_type (str): The type of activity to filter.
+            date (str): The end date for the date range.
+            num_days (int): The number of days to include in the date range.
+
+        Returns:
+            list: A list of pairs of duration and maximum pace values.
+        """
+
+        # Read the data from the file
+        data = self.data_source
+
+        # Convert date to datetime
+        end_date = pd.to_datetime(date)
+        start_date = end_date - pd.Timedelta(days=num_days)
+
+        # Convert 'Activity Date' column to datetime
+        data['Activity Date'] = pd.to_datetime(data['Activity Date'])
+
+        # Filter the activities based on the activity type and date range
+        filtered_activities = data[(data['Activity Type'] == activity_type) & 
+                                   (data['Activity Date'] >= start_date) & 
+                                   (data['Activity Date'] <= end_date)]
+
+        pace_curve = []
+        for duration in self.durations:
+            # Create a column name based on duration
+            col_name = f'Maximum Pace {self.format_duration(duration)}'
+
+            # Check if the column exists in the data
+            if col_name in filtered_activities.columns:
+                # Find the highest value for the 'Maximum Pace...' field for this duration
+                max_pace = filtered_activities[col_name].max()
+                pace_curve.append((duration, max_pace))
+            else:
+                # Handle cases where the column does not exist (e.g., append a default value or skip)
+                pace_curve.append((duration, None))  
+
+        return pace_curve
+
+
+    def plot_pace_curve(self, pace_curve, tested_cp):
+        """
+        Plots the pace curve.
+
+        Args:
+            pace_curve (list): A list of pairs of duration and maximum pace values.
+            tested_cp (float): The tested Critical Power.
+
+        Returns:
+            None
+        """
+
+        filtered_durations = []
+        filtered_paces = []
+
+        for duration, pace in pace_curve:
+            if pace > 0:
+                filtered_durations.append(self.format_duration(duration))
+                filtered_paces.append(pace)
+
+        # Convert durations to a numeric format
+        numeric_durations = np.arange(len(filtered_durations))
+
+        # Convert the list to a Pandas Series for easy rolling mean computation
+        pace_series = pd.Series(filtered_paces)
+
+        # Compute the moving average (rolling mean)
+        # The window size determines the smoothing level
+        smooth_paces = pace_series.rolling(window=2, center=True).mean().fillna(pace_series)
+
+        # Plotting the original and smoothed curve
```

Comment:
The pace curve is plotted by the method, and as it mainly deals with visualisation, the code appears perfect.